User guide for istioctl managing webhooks (#5162)
istio/istio.io: content/en/docs/tasks/security/webhook/index.md

---
title: Istio Webhook Management [Experimental]

> description: Shows how to manage webhooks in Istio through istioctl.

Reviewer: ~Shows~ How to manage webhooks in Istio through istioctl.

> description: Shows how to manage webhooks in Istio through istioctl.

Reviewer: ~Shows~ How to manage webhooks in Istio through istioctl.

Author: Revised accordingly.

weight: 100
keywords: [security,webhook]
---

{{< boilerplate experimental-feature-warning >}}

Istio has two webhooks: Galley and the sidecar injector. By default,
these webhooks manage their own configurations. From a

> security perspective, this default behavior is not recommended because a compromised webhook could then conduct

Reviewer: security perspective, this ~default~ behavior is not recommended ~because~ as it allows a compromised webhook ~could then~ to conduct

Author: The current one is more accurate.

> privilege escalation attacks.

Reviewer: privileged escalation attacks.

Author: It should be privilege escalation attacks.

> This task shows how to use `istioctl`, instead of the webhooks, to

Reviewer: This task shows how to use `istioctl`, instead of ~the~ webhooks, to

Author: Revised accordingly.

manage their configurations.

> ## Before you begin

Reviewer: ~## Before you begin~
## Getting started

Author: Revised accordingly.

* Install Istio with [DNS certificates configured](/docs/tasks/security/dns-cert) and
[`global.operatorManageWebhooks`]({{< github_file >}}/install/kubernetes/helm/istio/values.yaml) set to `true`.

* Install [`jq`](https://stedolan.github.io/jq/) for JSON parsing.

## Check webhook certificates

> To display the DNS names in the webhook certificates of Galley and the sidecar injector, we get the secret

Reviewer: To display the DNS names in the webhook certificates of Galley and the sidecar injector, ~we~ you need to get the secret

Author: Revised accordingly.

        rules:
          - operations:
            - CREATE
            - UPDATE
            apiGroups:
            - config.istio.io
            ... SKIPPED
        failurePolicy: Fail
        sideEffects: None
    {{< /text >}}

1.  Verify that there are no existing webhook configurations for Galley and the sidecar injector:

    {{< text bash >}}
    $ kubectl get mutatingwebhookconfiguration istio-sidecar-injector
    Error from server (NotFound): mutatingwebhookconfigurations.admissionregistration.k8s.io "istio-sidecar-injector" not found
    ~$ kubectl get validatingwebhookconfiguration istio-galley
    Error from server (NotFound): validatingwebhookconfigurations.admissionregistration.k8s.io "istio-galley" not found
    {{< /text >}}

1.  Use `istioctl` to enable the webhook configurations:

    {{< text bash >}}
    $ istioctl experimental post-install webhook enable --webhook-secret dns.istio-galley-service-account \
        --namespace istio-system --validation-path galley-webhook.yaml \
        --injection-path sidecar-injector-webhook.yaml
    {{< /text >}}

1.  To check that the sidecar injector webhook is working, verify that the webhook injects a

> sidecar container into an example pod with the following commands:

Reviewer: ~To~ Check that the sidecar injector webhook is working by verifying that the webhook injects a
sidecar container into an example pod with the following commands:

Author: The current one is more accurate.

    {{< text bash >}}
    $ kubectl create namespace test-injection; kubectl label namespaces test-injection istio-injection=enabled
    $ kubectl run --generator=run-pod/v1 --image=nginx nginx-app --port=80 -n test-injection
    $ kubectl get pod -n test-injection
    {{< /text >}}

>     The output from the `get pod` command should show the following output. The `2/2` value means that

Reviewer: The output from the `get pod` command should show the following ~output~. The `2/2` value means that

Author: Revised accordingly.

    {{< text bash >}}
    $ istioctl experimental post-install webhook status
    {{< /text >}}

1.  If you named the sidecar injector's configuration `istio-sidecar-injector`,
use the following command to show the configuration of the sidecar injector:

    {{< text bash >}}
    $ istioctl experimental post-install webhook status --validation=false --injection-config=istio-sidecar-injector
    {{< /text >}}

1.  If you named Galley's configuration `istio-galley`, show Galley's configuration with the following command:

    {{< text bash >}}
    $ istioctl experimental post-install webhook status --injection=false --validation-config=istio-galley
    {{< /text >}}

## Disable webhook configurations

1.  Disable Galley and the sidecar injector when using their default webhook configuration names:

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable
    {{< /text >}}

1.  If you named the sidecar injector's configuration `istio-sidecar-injector`,
disable the webhook with the following command:

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable --validation=false --injection-config=istio-sidecar-injector
    {{< /text >}}

> 1.  If you name Galleys's configuration `istio-galley`, disable the webhook with the following command:

Reviewer: If you named Galleys's configuration `istio-galley`, disable the webhook with the following command:

Author: Revised accordingly.

    {{< text bash >}}
    $ istioctl experimental post-install webhook disable --injection=false --validation-config=istio-galley
    {{< /text >}}

## Cleanup

> After completing this tutorial, you may delete the testing cluster created

Reviewer: ~After completing this tutorial, you may~ To delete the testing cluster created

at the beginning of this tutorial. You may also run the following command to delete

> the resources created.

Reviewer: ~After completing this tutorial, you may delete the testing cluster created
at the beginning of this tutorial.~ You ~may also~ can run the following command to delete
the resources created at the beginning of this tutorial.

Author: Revised accordingly.

---
title: Istio Webhook Management [Experimental]
description: Shows how to manage webhooks in Istio through istioctl.
weight: 100
keywords: [security,webhook]
---

{{< boilerplate experimental-feature-warning >}}

Istio has two webhooks: Galley and the sidecar injector. By default,
these webhooks manage their own configurations. From a
security perspective, this default behavior is not recommended because a compromised webhook could then conduct
privilege escalation attacks.

This task shows how to use `istioctl`, instead of the webhooks, to

> manage their configurations.

Reviewer 2: would be good to add a short explanation here to indicate why is this approach better than the default before getting into detailed steps.

Reviewer 3: @linsun We can explain the advantages of one approach over the other in a blog post. New readers of the task won't even know the old approach and having an explanation here will only confuse them.

Author: The explanation of why this approach is better is in the document as follows:
By default, these webhooks manage their own configurations. From a
security perspective, this default behavior is not recommended because a compromised webhook could then conduct privilege escalation attacks.

## Before you begin

> * Install Istio with [DNS certificates configured](/docs/tasks/security/dns-cert) and

Reviewer 2: is this correct?  getting 404 on https://preliminary.istio.io/docs/tasks/security/dns-cert

Reviewer 3: Great catch @linsun The correct path now is: /docs/tasks/traffic-management/ingress/ingress-certmgr/

Author: The 404 is because this PR refers to the guide in https://github.com/istio/istio.io/pull/5152, which has not been merged yet. https://github.com/istio/istio.io/pull/5152 should be merged first before merging this PR.

Reviewer 3: Thanks for that context @lei-tang

{{< text plain >}}
X509v3 Subject Alternative Name:
  DNS:istio-galley.istio-system.svc, DNS:istio-galley.istio-system
{{< /text >}}

{{< text plain >}}
X509v3 Subject Alternative Name:
  DNS:istio-sidecar-injector.istio-system.svc, DNS:istio-sidecar-injector.istio-system
{{< /text >}}

## Enable webhook configurations

1.  To generate the `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` configuration files, run the following
command. The [`values-istio-dns-cert.yaml`]({{< github_file >}}/install/kubernetes/helm/istio/example-values/values-istio-dns-cert.yaml) YAML file
contains an example DNS certificate configuration. Learn more in [the certificate guide](/docs/tasks/security/dns-cert).
The following command uses the following configurations:

    * The default Istio configuration.
    * The DNS certificate configuration in `values-istio-dns-cert.yaml`.

You can find more information in the `helm template` command [documentation](https://helm.sh/docs/helm/#helm-template).

    {{< text bash >}}
    $ helm template \
        --name=istio \
        --namespace=istio-system \

>         --values install/kubernetes/helm/istio/example-values/values-istio-dns-cert.yaml \

Reviewer 2: any equivalent istioctl cmd to generate the manifest?  I'm asking since the overall doc is depromoting helm.

Author: Done: changed to use istioctl to generate the manifest.

1.  Open the `istio-webhook-config.yaml` configuration file, search `'kind: ValidatingWebhookConfiguration'` and save
the `ValidatingWebhookConfiguration` of Galley to `galley-webhook.yaml`. The following
is a `ValidatingWebhookConfiguration` in an example `istio-webhook-config.yaml` (only
a part of the configuration is shown to save space).

    {{< text yaml >}}
    apiVersion: admissionregistration.k8s.io/v1beta1
    kind: ValidatingWebhookConfiguration
    metadata:
      name: istio-galley
      labels:
        app: galley
        chart: galley
        heritage: Tiller
        release: istio
        istio: galley
    webhooks:
      - name: pilot.validation.istio.io
        clientConfig:
          service:
            name: istio-galley
            namespace: istio-system
            path: "/admitpilot"
          caBundle: ""
        rules:
          - operations:
            - CREATE
            - UPDATE
            apiGroups:
            - config.istio.io
            ... SKIPPED
        failurePolicy: Fail
        sideEffects: None
    {{< /text >}}

1.  Use `istioctl` to enable the webhook configurations:

    {{< text bash >}}

>     $ istioctl experimental post-install webhook enable --validation --webhook-secret dns.istio-galley-service-account \

Reviewer 4: Is `--validation` necessary here?

Author: --validation can be skipped here since its default value is true. I have removed it from the command.

        --namespace istio-system --validation-path galley-webhook.yaml \
        --injection-path sidecar-injector-webhook.yaml
    {{< /text >}}

1.  To check that the sidecar injector webhook is working, verify that the webhook injects a
sidecar container into an example pod with the following commands:

    {{< text bash >}}
    $ kubectl create namespace test-injection; kubectl label namespaces test-injection istio-injection=enabled
    $ kubectl run --generator=run-pod/v1 --image=nginx nginx-app --port=80 -n test-injection
    $ kubectl get pod -n test-injection
    {{< /text >}}

    The output from the `get pod` command should show the following output. The `2/2` value means that
    the webhook injected a sidecar into the example pod:

    {{< text plain >}}
    NAME    READY   STATUS    RESTARTS   AGE
    nginx-app   2/2     Running   0          10s

>     {{< /text >}}

Reviewer 2: aren't this working without the steps here (e.g. using the default mode)?  how does user know if it is not using the default.

Reviewer 3: @linsun I don't understand you question here?

Author: Done: added a step to verify that there are no existing webhook configurations for Galley and the sidecar injector before running the webhook enable command.

1.  Check that the validation webhook is working:

    {{< text bash >}}
    $ kubectl create namespace test-validation
    $ kubectl apply -n test-validation -f - <<EOF
    apiVersion: networking.istio.io/v1alpha3
    kind: Gateway
    metadata:
      name: invalid-gateway
    spec:
      selector:
        # DO NOT CHANGE THESE LABELS
        # The ingressgateway is defined in install/kubernetes/helm/istio/values.yaml
        # with these labels
        istio: ingressgateway
    EOF
    {{< /text >}}

    The output from the gateway creation command should show the following output. The error
    in the output indicates that the validation webhook checked the gateway's configuration YAML file:

    {{< text plain >}}
    Error from server: error when creating "invalid-gateway.yaml": admission webhook "pilot.validation.istio.io" denied the request: configuration is invalid: gateway must have at least one server

>     {{< /text >}}

Reviewer 2: same comment here, i would expect this to fail in the default install also.

Reviewer 2: need more wording to explain why this is different and how to detect this is running in this new mode.

Reviewer 3: @linsun How this is different does not belong in the task. We can work on a blog post evaluating both approaches but let's try not to overburden readers with information here.

Author: Added a step to verify that there are no existing webhook configurations for Galley and the sidecar injector before running the webhook enable command.

> ---

Reviewer 5: Is this really a security topic? Seems more like something that belongs in the setup portion of the ops guide.

Author: Done: moved it to be under the content/en/docs/setup/install/webhook directory.

Author: Moved back to security based on the reviewer's comments in https://github.com/istio/istio.io/pull/5285.

title: Istio Webhook Management

> description: Shows how to manage webhooks in Istio through istioctl.

Reviewer 6: Can you include why a user would want to do this? Adding this doc, which is a lot of complicated and manual steps, makes it look like this is a required thing to user Istio properly

Author: Added the following explanation why a user would want to do this:
Istio has two webhooks (Galley and Sidecar Injector). In current Istio implementation,
Galley and Sidecar Injector manage their own webhook configurations, which from
security perspective is not recommended because a compromised webhook may conduct
privilege escalation attacks.

This task shows how to use `istioctl`, instead of Galley and Sidecar Injector, to
manage the webhook configurations of Galley and Sidecar Injector.

weight: 10
keywords: [security,webhook]
---

> {{< warning >}}

Reviewer 5: Use the boilerplate for this warning please.

Author: Done.

---
title: Istio Webhook Management [Experimental]
description: How to manage webhooks in Istio through istioctl.
weight: 100
keywords: [security,webhook]
---

{{< boilerplate experimental-feature-warning >}}

Istio has two webhooks: Galley and the sidecar injector. By default,
these webhooks manage their own configurations. From a
security perspective, this default behavior is not recommended because a compromised webhook could then conduct
privilege escalation attacks.

> This task shows how to use `istioctl`, instead of webhooks, to

Reviewer 7: ```suggestion
This task shows how to use the new [{{< istioctl >}} x post-install webhook](/docs/reference/commands/istioctl/#istioctl-experimental-post-install-webhook) command to
```

Author: Revised accordingly.

> manage their configurations.

Reviewer 7: ```suggestion
securely manage the configurations of the webhooks.
```

Author: Revised accordingly.

{{< text bash >}}
$ kubectl get secret dns.istio-galley-service-account -n istio-system -o json | jq -r '.data["cert-chain.pem"]' | base64 --decode | openssl x509 -in - -text -noout
$ kubectl get secret dns.istio-sidecar-injector-service-account -n istio-system -o json | jq -r '.data["cert-chain.pem"]' | base64 --decode | openssl x509 -in - -text -noout
{{< /text >}}

The output from the above commands should include the DNS names of Galley and the sidecar injector, respectively:

{{< text plain >}}
X509v3 Subject Alternative Name:
  DNS:istio-galley.istio-system.svc, DNS:istio-galley.istio-system
{{< /text >}}

{{< text plain >}}
X509v3 Subject Alternative Name:
  DNS:istio-sidecar-injector.istio-system.svc, DNS:istio-sidecar-injector.istio-system
{{< /text >}}

## Enable webhook configurations

1.  To generate the `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` configuration files, run the following
command.

    {{< text bash >}}
    $ istioctl manifest generate > istio.yaml
    {{< /text >}}

<!-- TODO (lei-tang): improve the UX for obtain MutatingWebhookConfiguration -->

> 1.  Open the `istio.yaml` configuration file, search `'kind: MutatingWebhookConfiguration'` and save

Reviewer 7: ```suggestion
1.  Open the `istio.yaml` configuration file, search for `'kind: MutatingWebhookConfiguration'` and save
```

Author: Done.

the `MutatingWebhookConfiguration` of the sidecar injector to `sidecar-injector-webhook.yaml`. The following
is a `MutatingWebhookConfiguration` in an example `istio.yaml`.

    {{< text yaml >}}
    apiVersion: admissionregistration.k8s.io/v1beta1
    kind: MutatingWebhookConfiguration
    metadata:
      name: istio-sidecar-injector
      labels:
        app: sidecarInjectorWebhook
        release: istio
    webhooks:
      - name: sidecar-injector.istio.io
        clientConfig:
          service:
            name: istio-sidecar-injector
            namespace: istio-system
            path: "/inject"
          caBundle: ""
        rules:
          - operations: [ "CREATE" ]
            apiGroups: [""]
            apiVersions: ["v1"]
            resources: ["pods"]
        failurePolicy: Fail
        namespaceSelector:
          matchLabels:
            istio-injection: enabled
    {{< /text >}}

<!-- TODO (lei-tang): improve the UX for obtain ValidatingWebhookConfiguration -->

> 1.  Open the `istio.yaml` configuration file, search `'kind: ValidatingWebhookConfiguration'` and save

Reviewer 7: ```suggestion
1.  Open the `istio.yaml` configuration file, search for `kind: ValidatingWebhookConfiguration` and save
```

Author: Done.

            apiGroups:
            - config.istio.io
            ... SKIPPED
        failurePolicy: Fail
        sideEffects: None
    {{< /text >}}

1.  Verify that there are no existing webhook configurations for Galley and the sidecar injector:

    {{< text bash >}}
    $ kubectl get mutatingwebhookconfiguration istio-sidecar-injector
    Error from server (NotFound): mutatingwebhookconfigurations.admissionregistration.k8s.io "istio-sidecar-injector" not found
    ~$ kubectl get validatingwebhookconfiguration istio-galley
    Error from server (NotFound): validatingwebhookconfigurations.admissionregistration.k8s.io "istio-galley" not found
    {{< /text >}}

1.  Use `istioctl` to enable the webhook configurations:

    {{< text bash >}}
    $ istioctl experimental post-install webhook enable --webhook-secret dns.istio-galley-service-account \
        --namespace istio-system --validation-path galley-webhook.yaml \
        --injection-path sidecar-injector-webhook.yaml
    {{< /text >}}

1.  To check that the sidecar injector webhook is working, verify that the webhook injects a
sidecar container into an example pod with the following commands:

    {{< text bash >}}

>     $ kubectl create namespace test-injection; kubectl label namespaces test-injection istio-injection=enabled

Reviewer 7: ```suggestion
    $ kubectl create namespace test-injection
    $ kubectl label namespaces test-injection istio-injection=enabled
```

Author: Done.

    $ kubectl run --generator=run-pod/v1 --image=nginx nginx-app --port=80 -n test-injection
    $ kubectl get pod -n test-injection
    {{< /text >}}

    The output from the `get pod` command should show the following. The `2/2` value means that
    the webhook injected a sidecar into the example pod:

    {{< text plain >}}

>     NAME    READY   STATUS    RESTARTS   AGE

Reviewer 7: ```suggestion
    NAME        READY   STATUS    RESTARTS   AGE
```

Author: Done.

from Kubernetes, parse it, decode it, and view the text output with the following commands:

{{< text bash >}}
$ kubectl get secret dns.istio-galley-service-account -n istio-system -o json | jq -r '.data["cert-chain.pem"]' | base64 --decode | openssl x509 -in - -text -noout
$ kubectl get secret dns.istio-sidecar-injector-service-account -n istio-system -o json | jq -r '.data["cert-chain.pem"]' | base64 --decode | openssl x509 -in - -text -noout
{{< /text >}}

The output from the above commands should include the DNS names of Galley and the sidecar injector, respectively:

{{< text plain >}}
X509v3 Subject Alternative Name:
  DNS:istio-galley.istio-system.svc, DNS:istio-galley.istio-system
{{< /text >}}

{{< text plain >}}
X509v3 Subject Alternative Name:
  DNS:istio-sidecar-injector.istio-system.svc, DNS:istio-sidecar-injector.istio-system
{{< /text >}}

## Enable webhook configurations

1.  To generate the `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` configuration files, run the following
command.

    {{< text bash >}}
    $ istioctl manifest generate > istio.yaml
    {{< /text >}}

> <!-- TODO (lei-tang): improve the UX for obtain MutatingWebhookConfiguration -->

Reviewer 7: Remove this line. It is causing the numbered list not to render properly.

Author: Done, thanks!

1.  Open the `istio.yaml` configuration file, search for `kind: MutatingWebhookConfiguration` and save
the `MutatingWebhookConfiguration` of the sidecar injector to `sidecar-injector-webhook.yaml`. The following
is a `MutatingWebhookConfiguration` in an example `istio.yaml`.

    {{< text yaml >}}
    apiVersion: admissionregistration.k8s.io/v1beta1
    kind: MutatingWebhookConfiguration
    metadata:
      name: istio-sidecar-injector
      labels:
        app: sidecarInjectorWebhook
        release: istio
    webhooks:
      - name: sidecar-injector.istio.io
        clientConfig:
          service:
            name: istio-sidecar-injector
            namespace: istio-system
            path: "/inject"
          caBundle: ""
        rules:
          - operations: [ "CREATE" ]
            apiGroups: [""]
            apiVersions: ["v1"]
            resources: ["pods"]
        failurePolicy: Fail
        namespaceSelector:
          matchLabels:
            istio-injection: enabled
    {{< /text >}}

> <!-- TODO (lei-tang): improve the UX for obtain ValidatingWebhookConfiguration -->

Reviewer 7: Same as above.

Author: Done, thanks!

---
title: Istio Webhook Management
description: Shows how to manage webhooks in Istio through istioctl.
weight: 10
keywords: [security,webhook]
---

{{< warning >}}
The following information describes an experimental feature, which is intended
for evaluation purposes only.
{{< /warning >}}

> This task shows how to how to manage webhooks in Istio through `istioctl`.

Reviewer 6: This doesn't really explain why I would ever want to do this?

Author: Done: added an explanation of why want to do this.

## Before you begin

> * Create a new Kubernetes cluster to run the example in this tutorial.

Reviewer 6: nit: I don't think this is needed

Author: Creating a new cluster is not strictly needed, but to on the safe side -> avoid affecting an existing cluster.

    $ kubectl get secret dns.istio-sidecar-injector-service-account -n istio-system -o json | jq -r '.data["cert-chain.pem"]' | base64 --decode | openssl x509 -in - -text -noout
    {{< /text >}}

    The output from the above commands should include the DNS names of Galley and Sidecar Injector, respectively:

    {{< text plain >}}
    X509v3 Subject Alternative Name:
      DNS:istio-galley.istio-system.svc, DNS:istio-galley.istio-system
    {{< /text >}}

    {{< text plain >}}
    X509v3 Subject Alternative Name:
      DNS:istio-sidecar-injector.istio-system.svc, DNS:istio-sidecar-injector.istio-system
    {{< /text >}}

## `istioctl` enables webhook configurations

1.  Generate `MutatingWebhookConfiguration` and `ValidatingWebhookConfiguration` by running the following
command:

    {{< text bash >}}
    $ helm template \
        --name=istio \
        --namespace=istio-system \
        --values install/kubernetes/helm/istio/example-values/values-istio-dns-cert.yaml \
        install/kubernetes/helm/istio > istio-webhook-config.yaml
    {{< /text >}}

> 1.  From `istio-webhook-config.yaml`, search `'kind: MutatingWebhookConfiguration'` and save

Reviewer 6: If this is for more then experimental we need to improve the UX here. If its is just experimental we should make that more clear

Author: Added TODOs to improve the UX here. The document has a warning at the top saying this is an experimental feature.

            - networking.istio.io
            apiVersions:
            - "*"
            resources:
            - destinationrules
            - envoyfilters
            - gateways
            - serviceentries
            - sidecars
            - virtualservices
        failurePolicy: Fail
        sideEffects: None
      - name: mixer.validation.istio.io
        clientConfig:
          service:
            name: istio-galley
            namespace: istio-system
            path: "/admitmixer"
          caBundle: ""
        rules:
          - operations:
            - CREATE
            - UPDATE
            apiGroups:
            - config.istio.io
            apiVersions:
            - v1alpha2
            resources:
            - rules
            - attributemanifests
            - circonuses
            - deniers
            - fluentds
            - kubernetesenvs
            - listcheckers
            - memquotas
            - noops
            - opas
            - prometheuses
            - rbacs
            - solarwindses
            - stackdrivers
            - cloudwatches
            - dogstatsds
            - statsds
            - stdios
            - apikeys

>             - authorizations

Reviewer 6: nit: probably don't need to paste the full config?

Author: Done: skipped most of the config.
